builtin/logical/aws/rotation.go

```diff
@@ -0,0 +1,125 @@
+package aws
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/vault/sdk/logical"
+	"github.com/hashicorp/vault/sdk/queue"
+)
+
+func (b *backend) initQueue(_ context.Context, _ *logical.InitializationRequest) error {
+	return nil
+}
+
+func (b *backend) rotateExpiredStaticCreds(ctx context.Context, req *logical.Request) error {
+	var errs *multierror.Error
+
+	for {
+		keepGoing, err := b.rotateCredential(ctx, req.Storage)
+		if err != nil {
+			errs = multierror.Append(errs, err)
+		}
+		if !keepGoing {
+			if errs != nil && len(errs.Errors) != 0 {
+				return fmt.Errorf("error(s) occurred while rotating expired static credentials: %w", errs)
+			} else {
+				return nil
+			}
+		}
+	}
+}
+
+func (b *backend) rotateCredential(ctx context.Context, storage logical.Storage) (bool, error) {
+	// If queue is empty or first item does not need a rotation (priority is next rotation timestamp) there is nothing to do
+	item, err := b.credRotationQueue.Pop()
+	if err != nil && err == queue.ErrEmpty {
+		return false, nil
+	}
+	if item.Priority > time.Now().Unix() {
+		// no rotation required
+		// push the item back into priority queue
+		err = b.credRotationQueue.Push(item)
+		if err != nil {
+			return false, fmt.Errorf("failed to add item into the rotation queue for role %q: %w", item.Key, err)
+		}
+		return false, nil
+	}
+
+	cfg := item.Value.(staticRoleConfig)
+
+	err = b.createCredential(ctx, storage, cfg)
+	if err != nil {
+		return false, err
+	}
+
+	// set new priority and re-queue
+	item.Priority = time.Now().Add(cfg.RotationPeriod).Unix()
+	err = b.credRotationQueue.Push(item)
+	if err != nil {
+		return false, fmt.Errorf("failed to add item into the rotation queue for role %q: %w", cfg.Name, err)
+	}
+
+	return true, nil
+}
+
+func (b *backend) createCredential(ctx context.Context, storage logical.Storage, cfg staticRoleConfig) error {
+	iamClient, err := b.clientIAM(ctx, storage)
+	if err != nil {
+		return fmt.Errorf("unable to get the AWS IAM client: %w", err)
+	}
+
+	// IAM users can have at most 2 set of keys at a time. If 2 sets already exists we must delete the oldest one to be
+	// able to create a new set of credentials.
```

Comment:
I wonder if we can count on this always holding true. If not, maybe a safer approach would be to iterate over all of the keys when there are two or more, and delete the oldest one.

Reply:
I think switching to iterating and deleting the oldest is a better solution, but I also want to mention the limitations of the fix.

As far as I know, the only way for us to know the maximum number of allowed keys is via AWS documentation (i.e., it's not exposed in code). Switching to iteration would guarantee that we always delete the oldest, even if Amazon increases the quota, but we're still relying on a hardcoded value to know we're only deleting keys if necessary.